### ISSUE TITLE
Remote Attestation Service Provider Sample Crypto Library to Real Crypto Library

### ISSUE THREAD

**Issue by JoshLind** [intel/linux-sgx#58]
TL;DR:
1) Should we just be able to swap out the lib_sample_libcrypto library for the sgx_tcrypto library?
2) If we modified the lib_sample_libcrypto library to use real random numbers instead of deterministically returning 9 each time, could it be used in a production environment?

Hi there,

I'm playing around with the Remote Attestation code example and see that the service provider uses a sample cryptographic library for debugging purposes (as described on page 82 of the developer guide[1]). This all makes sense to me, and I'm now trying to modify the example so that the service provider code also executes in an enclave. (The idea being that I could eventually remotely attest between two enclaves using this code -- ignoring any network communication outside the enclave etc.) 

Of course, I can't use the sample_libcrypto inside an enclave for the service provider and in a production enclave I should instead use the real crypto library (sgx_tcrypto). It seems to me that by modifying the Service Provider/ECP/IAS_RA code, I should just be able to change all of the references from the sample crypto functions/data structures to the sgx crypto functions/data structures (for example, change sample_rijndael128_cmac_msg(...) -> sgx_rijndael128_cmac_msg(...), sample_status_t -> sgx_status_t  and so on.) Is this the idea?

I have tried this, but it doesn't seem to work (once I've made all the changes and updated the edl file/ memory allocations so that messages are appropriately passed across the enclave boundary, the sample Remote Attestation code fails when the service provider tries to verify the cmac of the message, after calling sgx_rijndael128_cmac_msg (used to be sample_rijndael128_cmac_msg), the code fails the check here [3].  It seems to me like the example is structured in a way that the two libraries should be replaceable, unless there are differences between the data structures? I'm wondering if this was the original intention of the example?

Looking at the source code for the sample crypto library[2] it looks like the reason this library shouldn't be used in a production environment is because the __do_get_rand_32 function is deterministic (for debugging). Is this the only reason? If so, could one replace this function to produce real random numbers and thus use the library in a production environment?

Thanks so much for your help!!
Josh

[1] https://download.01.org/intel-sgx/linux-1.7/docs/Intel_SGX_SDK_Developer_Reference_Linux_1.7_Open_Source.pdf

[2] https://github.com/01org/linux-sgx/tree/master/sdk/sample_libcrypto

[3]https://github.com/01org/linux-sgx/blob/master/SampleCode/RemoteAttestation/service_provider/service_provider.cpp#L496

**Comment by JoshLind** [comment]
No worries -- I managed to successfully swap lib_sample_libcrypto for the sgx_tcrypto inside the enclave :)

Thanks anyway!
Josh

**Comment by JoshLind** [comment]
Reopening this to see if I can get any comments/insight on the approach here :)

**Comment by andyzyb** [comment]
I think the approach to replace simple_libcrypto with sgx_tcrypto should work. But please do all your security evaluation for the code as it is provided just as reference code.
